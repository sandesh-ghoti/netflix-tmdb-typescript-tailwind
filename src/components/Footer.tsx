import * as React from "react";

interface IFooterProps {}

const Footer: React.FunctionComponent<IFooterProps> = () => {
  return <></>;
};

export default Footer;
