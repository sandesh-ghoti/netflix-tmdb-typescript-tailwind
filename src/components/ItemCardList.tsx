import * as React from "react";
import { Movie, Person, TV } from "tmdb-ts";
import { PosterCard } from "./PosterCard";
interface CardProps {
  list: Movie[] | TV[] | Person[];
  title: string;
}

export const ItemCardList: React.FunctionComponent<CardProps> = (props) => {
  return (
    <section className="w-full bg-white dark:bg-gray-800 dark:border-gray-700">
      <h2 className="my-4 font-bold text-4xl text-center text-black dark:text-white">
        {props.title}
      </h2>
      <ul className="grid grid-cols-2 md:grid-cols-3 lg:grid-cols-4 xl:grid-cols-5 2xl:grid-cols-6 gap-4">
        {props.list.map((item) => (
          <li key={item.id}>
            {isMovie(item) ? (
              <PosterCard
                id={item.id}
                poster_path={item.poster_path}
                media_type="movie"
                title={item.title}
                vote_average={item.vote_average}
              />
            ) : isTV(item) ? (
              <PosterCard
                id={item.id}
                poster_path={item.poster_path}
                media_type="tv"
                title={item.name} // TV shows use the 'name' property instead of 'title'
                vote_average={item.vote_average}
              />
            ) : (
              <PosterCard
                id={item.id}
                poster_path={item.profile_path}
                media_type="person"
                title={item.name}
                vote_average={item.popularity}
              />
            )}
          </li>
        ))}
      </ul>
    </section>
  );
};

export function isMovie(item: Movie | TV | Person): item is Movie {
  return (item as Movie).title !== undefined;
}

export function isTV(item: Movie | TV | Person): item is TV {
  return (item as TV).name !== undefined;
}

export function isPerson(item: Movie | TV | Person): item is Person {
  return (item as Person).name !== undefined;
}
