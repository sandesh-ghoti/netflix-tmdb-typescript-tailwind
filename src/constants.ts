export const TMDB_API_KEY = process.env.TMDB_API_KEY;
export const TMDB_ACCESS_TOKEN = process.env.TMDB_ACCESS_TOKEN;
export const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
export const BASE_URL = "https://api.themoviedb.org/3";
export const IMG_CDN_URL = "https://image.tmdb.org/t/p/w500";
export const YT_THUMBNAIL_URL = "https://i.ytimg.com/vi/";
//https://i.ytimg.com/vi/SzINZZ6iqxY/hqdefault.jpg
export const FIREBASE_API_KEY = process.env.FIREBASE_API_KEY;
export const FIREBASE_AUTH_DOMAIN = process.env.FIREBASE_AUTH_DOMAIN;
export const FIREBASE_PROJECT_ID = process.env.FIREBASE_PROJECT_ID;
export const FIREBASE_STORAGE_BUCKET = process.env.FIREBASE_STORAGE_BUCKET;
export const FIREBASE_MESSAGING_SENDER_ID =
  process.env.FIREBASE_MESSAGING_SENDER_ID;
export const FIREBASE_APP_ID = process.env.FIREBASE_APP_ID;
export const FIREBASE_MEASUREMENT_ID = process.env.FIREBASE_MEASUREMENT_ID;
export const IMG_CONFIG = {
  base_url: "http://image.tmdb.org/t/p/",
  secure_base_url: "https://image.tmdb.org/t/p/",
  backdrop_sizes: ["w300", "w780", "w1280", "original"],
  logo_sizes: ["w45", "w92", "w154", "w185", "w300", "w500", "original"],
  poster_sizes: ["w92", "w154", "w185", "w342", "w500", "w780", "original"],
  profile_sizes: ["w45", "w185", "h632", "original"],
  still_sizes: ["w92", "w185", "w300", "original"],
};
export const SUPPORTED_LANGUAGES = [
  { identifier: "en-Us", name: "English" },
  { identifier: "hn", name: "Hindi" },
  { identifier: "es", name: "Spanish" },
];
