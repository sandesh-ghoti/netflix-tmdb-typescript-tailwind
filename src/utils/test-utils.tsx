import { cleanup, render, RenderOptions } from "@testing-library/react";
import React, { PropsWithChildren } from "react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import { setupStore } from "../store";
import type { AppStore, RootState } from "../store";

afterEach(() => {
  cleanup();
});
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
  initialEntries?: string[];
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    initialEntries = ["/"],
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <MemoryRouter initialEntries={initialEntries}>{children}</MemoryRouter>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithProvidersWithoutRouter(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
