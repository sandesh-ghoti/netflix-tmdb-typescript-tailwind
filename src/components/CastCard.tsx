import * as React from "react";
import { IMG_CONFIG } from "../constants";
import { useNavigate } from "react-router-dom";

export interface ICastCardProps {
  profile_path: string;
  heading: string;
  subHeading: string;
  redirect?: string | null;
  id: number | string;
}

export const CastCard: React.FunctionComponent<ICastCardProps> = (data) => {
  if (!data) return null;
  const navigate = useNavigate();
  const handleRedirect = () => {
    if (data.redirect) {
      navigate(`${data.redirect}/${data.id}`);
    }
  };
  return (
    <div
      className=" flex flex-col w-full h-full justify-center overflow-hidden bg-slate-700 bg-opacity-25 rounded-lg shadow-md shadow-gray-700"
      onClick={handleRedirect}
    >
      <img
        src={`${IMG_CONFIG.secure_base_url}${IMG_CONFIG.profile_sizes[2]}${data.profile_path}`}
        alt={data?.heading}
        className="w-full h-4/5 object-cover"
      />
      <div className="w-full mx-auto my-auto py-2 flex flex-col justify-center items-center">
        <h1 className=" text-center text-sm font-semibold">{data?.heading}</h1>
        <div className="text-xs text-wrap text-center">{data?.subHeading}</div>
      </div>
    </div>
  );
};
