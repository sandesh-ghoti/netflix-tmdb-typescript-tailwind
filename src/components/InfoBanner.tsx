import * as React from "react";
import { MovieDetails } from "tmdb-ts";
import { IMG_CONFIG } from "../constants";
import { Ratings } from "./Ratings";

interface IInfoBannerProps {
  data: MovieDetails;
}

const InfoBanner: React.FunctionComponent<IInfoBannerProps> = ({ data }) => {
  return (
    <section
      style={
        data?.backdrop_path
          ? {
              width: "100%",
              height: "100%",
              backgroundImage: `url(${IMG_CONFIG.secure_base_url}${IMG_CONFIG.backdrop_sizes[3]}${data?.backdrop_path})`,
              backgroundSize: "cover",
              backgroundPosition: "center",
              backgroundRepeat: "no-repeat",
            }
          : { backgroundColor: "bg-slate-800" }
      }
    >
      <div
        className={
          " w-full h-full mx-auto bg-slate-900 bg-opacity-70 grid grid-cols-3 xl:grid-cols-4 gap-4 items-center overflow-hidden p-8 mt-6"
        }
      >
        <img
          src={`${IMG_CONFIG.secure_base_url}${IMG_CONFIG.poster_sizes[4]}${data?.poster_path}`}
          alt={`${data?.title}`}
          className=" col-span-1 rounded-md shadow-md shadow-slate-400 my-auto"
        />
        <div className="col-span-2 xl:col-span-3 my-auto flex flex-col">
          <h1 className="text-xl font-bold text-slate-100 my-2 md:text-2xl lg:text-3xl">
            {data?.title}
            {" ("}
            {data.release_date.split("-")[0]}
            {")"}
          </h1>
          <span className="text-sm mb-4 font-serif italic">
            {data?.tagline}
          </span>
          <div className="my-1">
            <Ratings rating={parseFloat((data?.vote_average / 2).toFixed(1))} />
          </div>
          <div className="flex flex-row flex-wrap gap-2 max-md:text-sm max-md:gap-1 my-1">
            <h4 className="font-bold text-slate-100">Total Votes</h4>:{" "}
            <span>{data?.vote_count}</span>
          </div>
          <div className="flex flex-row flex-wrap gap-2 max-md:text-sm max-md:gap-1 my-1">
            <h4 className="font-bold text-slate-100">Geners</h4>:{" "}
            <span>{data.genres.map((genre) => genre.name).join(", ")}</span>
          </div>
          <div className="flex flex-row flex-wrap gap-2 max-md:text-sm max-md:gap-1 my-1">
            <h4 className="font-bold text-slate-100">Runtime</h4>:{" "}
            <span>{data?.runtime} minutes</span>
          </div>
          <div className="flex flex-row flex-wrap gap-2 max-md:text-sm max-md:gap-1 my-1">
            <h4 className="font-bold text-slate-100">Release Date</h4>:{" "}
            <span>{data?.release_date}</span>
          </div>
          <div className="flex flex-row flex-wrap gap-2 max-md:text-sm max-md:gap-1 my-1">
            <h4 className="font-bold text-slate-100">Status</h4>:{" "}
            <span>{data?.status}</span>
          </div>
          <p className="text-sm text-zinc-200 my-2">{data?.overview}</p>
        </div>
      </div>
    </section>
  );
};

export default InfoBanner;
