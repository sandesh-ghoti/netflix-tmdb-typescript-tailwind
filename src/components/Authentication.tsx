import React, { useState } from "react";
import { TMDB_API_KEY } from "../constants";
import {
  loginFailure,
  loginIntialise,
  loginSuccess,
  useAppDispatch,
  useGetRequestTokenQuery,
  useSetSessionTokenMutation,
} from "../store";
import { ErrorResponse, ValidateLoginResponse } from "../utils/commonTypes";
import { useLocation, useNavigate } from "react-router-dom";
const LoginComponent: React.FC = () => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const location = useLocation();

  const {
    data: requestToken,
    isLoading: isRequestTokenLoading,
    refetch,
  } = useGetRequestTokenQuery();
  const [setSessionId, { data: sessionData }] = useSetSessionTokenMutation();
  const handleLogin = async () => {
    setError(null);
    setLoading(true);
    const from = location.state?.from?.pathname || "/";
    try {
      dispatch(loginIntialise());
      if (!requestToken?.request_token) {
        throw new Error("Failed to fetch request token");
      }
      if (new Date().getTime() > new Date(requestToken.expires_at).getTime()) {
        console.log("request token expired");
        await refetch();
      }
      const response = await fetch(
        `https://api.themoviedb.org/3/authentication/token/validate_with_login?api_key=${TMDB_API_KEY}`,
        {
          method: "POST",
          body: JSON.stringify({
            username,
            password,
            request_token: requestToken?.request_token,
          }),
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      const data: ValidateLoginResponse = await response.json();
      if (!data.success) {
        const errorResponse = data as unknown as ErrorResponse;
        throw new Error(
          `Invalid username or password ${errorResponse.status_message}`
        );
      }
      const sessionResponse = await setSessionId({
        request_token: requestToken?.request_token,
      });
      console.log(
        "sessionResponse:",
        sessionResponse,
        "sessionData:",
        sessionData
      );
      if (!sessionResponse.data?.success) {
        throw new Error("Failed to set session token");
      }
      dispatch(loginSuccess(sessionResponse.data.session_id));
      setLoading(false);
      navigate(from, { replace: true });
    } catch (error) {
      dispatch(loginFailure((error as Error).message));
      setError((error as Error).message);
      console.error("Login failed:", (error as Error).message);
    }
    setLoading(false);
  };

  return (
    <div className="flex flex-col items-center justify-center h-screen">
      <h2 className="text-2xl font-bold mb-4">Login</h2>
      <div className="w-64">
        <input
          type="text"
          value={username}
          onChange={(e) => setUsername(e.target.value)}
          placeholder="Username"
          className="w-full px-4 py-2 mb-4 rounded border bg-gray-100 dark:bg-gray-700 focus:outline-none"
        />
        <input
          type="password"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
          placeholder="Password"
          className="w-full px-4 py-2 mb-4 rounded border bg-gray-100 dark:bg-gray-700 focus:outline-none"
        />
        {error && <p className="text-red-500 text-sm mb-4">{error}</p>}
        <button
          onClick={handleLogin}
          className={`w-full bg-blue-500 text-white font-bold py-2 px-4 rounded focus:outline-none ${
            loading ? "opacity-50 cursor-not-allowed" : ""
          }`}
          disabled={loading || isRequestTokenLoading}
        >
          {isRequestTokenLoading
            ? "Request Token Loading..."
            : loading
            ? "Logging in..."
            : "Login"}
        </button>
      </div>
    </div>
  );
};

export default LoginComponent;
