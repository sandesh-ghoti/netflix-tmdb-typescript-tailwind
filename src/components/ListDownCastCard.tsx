import * as React from "react";
import { CastCard, ICastCardProps } from "./CastCard";

interface IListDownCastCardProps {
  list: ICastCardProps[];
}

export const ListDownCastCard: React.FunctionComponent<
  IListDownCastCardProps
> = (props) => {
  return (
    <ul className="flex flex-row overflow-scroll w-full gap-4 sm:gap-6 md:gap-8 lg:gap-10 p-4">
      {props.list.map((item) => (
        <li
          key={item.id}
          className="flex-none w-1/2 sm:w-1/3 md:w-1/4 lg:w-1/5 xl:w-1/6"
        >
          <CastCard {...item} />
        </li>
      ))}
    </ul>
  );
};
