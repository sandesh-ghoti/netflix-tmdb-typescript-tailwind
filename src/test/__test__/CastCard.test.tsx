import { screen } from "@testing-library/react";
import { CastCard } from "../../components/CastCard";
import { renderWithProviders } from "../../utils/test-utils";
import { creditList } from "../mocks/dummyMockData";
import { IMG_CONFIG } from "../../constants";
const data = {
  profile_path: creditList.cast[0].profile_path,
  heading: creditList.cast[0].name,
  subHeading: creditList.cast[0].character,
  redirect: `people/${creditList.cast[0].id}`,
  id: creditList.cast[0].id,
};
describe("CastCard", async () => {
  test("app renders correctly", async () => {
    const castCard = renderWithProviders(<CastCard {...data} />);
    expect(castCard.container).toBeTruthy();
    expect(castCard.container.querySelector("h1")?.textContent).toBe(
      data.heading
    );
    const card = castCard.container.querySelector("div") as HTMLDivElement;
    expect(card).toBeInTheDocument();
  });
  it("renders correctly with the provided data", async () => {
    renderWithProviders(<CastCard {...data} />);
    screen.debug();

    const image = screen.getByRole("img");
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute(
      "src",
      `${IMG_CONFIG.secure_base_url}${IMG_CONFIG.profile_sizes[2]}${data.profile_path}`
    );

    const heading = screen.getByText(data.heading);
    expect(heading).toBeInTheDocument();

    const subHeading = screen.getByText(data.subHeading);
    expect(subHeading).toBeInTheDocument();
  });
});
