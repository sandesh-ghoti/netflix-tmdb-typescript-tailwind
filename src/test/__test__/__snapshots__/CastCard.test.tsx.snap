// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CastCard > test snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=" flex flex-col w-full h-full justify-center overflow-hidden bg-slate-700 bg-opacity-25 rounded-lg shadow-md shadow-gray-700"
      >
        <img
          alt="Owen Teague"
          class="w-full h-4/5 object-cover"
          src="https://image.tmdb.org/t/p/h632/tgCkGE0LIggyjMmgSwHhpZAkfJs.jpg"
        />
        <div
          class="w-full mx-auto my-auto py-2 flex flex-col justify-center items-center"
        >
          <h1
            class=" text-center text-sm font-semibold"
          >
            Owen Teague
          </h1>
          <div
            class="text-xs text-wrap text-center"
          >
            Noa
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=" flex flex-col w-full h-full justify-center overflow-hidden bg-slate-700 bg-opacity-25 rounded-lg shadow-md shadow-gray-700"
    >
      <img
        alt="Owen Teague"
        class="w-full h-4/5 object-cover"
        src="https://image.tmdb.org/t/p/h632/tgCkGE0LIggyjMmgSwHhpZAkfJs.jpg"
      />
      <div
        class="w-full mx-auto my-auto py-2 flex flex-col justify-center items-center"
      >
        <h1
          class=" text-center text-sm font-semibold"
        >
          Owen Teague
        </h1>
        <div
          class="text-xs text-wrap text-center"
        >
          Noa
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
